### PR TITLE
Interpolation fields factory

### PR DESCRIPTION
**Pull request by halungge**
Set up a field source for interpolation fields.

(Restrictions) Adds a `FieldProvider` implementation for direct call of `gtx.field_operator` (where possible), this one currently duplicates some code with from the implementation for the `gtx.program.` It also allows for computing on different backend than the target backend.

### CODE REVIEW

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
cscs-ci run default

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 1]
```diff
@@ -275,26 +280,31 @@ def int_variable(
...
         _log.debug(f"reading {name}: transposing = {transpose}")
-        data = self.variable(name, indices, dtype=gtx.int32)
-        return np.transpose(data) if transpose else data
+        return self.variable(name, indices, transpose=transpose, dtype=gtx.int32)
 
     def variable(
-        self, name: FieldName, indices: np.ndarray = None, dtype: np.dtype = gtx.float64
+        self,
+        name: FieldName,
+        indices: np.ndarray = None,
```
Comment (nfarabullini) [inline model/common/src/icon4py/model/common/grid/grid_manager.py:288]:
```suggestion
        indices: field_alloc.NDArray = None,
```

File: model/common/src/icon4py/model/common/states/factory.py [hunk 2]
```diff
@@ -177,15 +191,44 @@ def get(
...
+        return name in self._sources._providers or name in self._sources.metadata.keys()
+
     def register_provider(self, provider: FieldProvider):
+        # dependencies must be provider by this field source or registered in sources
         for dependency in provider.dependencies:
-            if dependency not in self.providers.keys():
+            if not (dependency in self._providers.keys() or self._provided_by_source(dependency)):
                 raise ValueError(
-                    f"Dependency '{dependency}' not found in registered providers of source {self.__class__}"
+                    f"Missing dependency: '{dependency}' not found in registered of sources {self.__class__}"
```
Comment (nfarabullini) [inline model/common/src/icon4py/model/common/states/factory.py:202]:
```suggestion
                    f"Missing dependency: '{dependency}' in registered of sources {self.__class__}"
```

File: model/common/src/icon4py/model/common/states/factory.py [hunk 3]
```diff
@@ -213,6 +256,130 @@ def func(self) -> Callable:
...
+        deps: dict[str, str],  # keyword arg to (field_operator, field_name) need: src
+        params: Optional[
+            dict[str, state_utils.ScalarType]
+        ] = None,  # keyword arg to (field_operator, field_name)
+    ):
+        self._func = func
+        self._dims = domain
+        self._dependencies = deps
+        self._output = fields
+        self._params = params if params is not None else {}
```
Comment (nfarabullini) [inline model/common/src/icon4py/model/common/states/factory.py:284]:
```suggestion
        self._params = {} if params is None else params
```
just as a personal preference

File: model/common/tests/states_test/test_factory.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    test_source = SimpleSource(data_=data, grid=grid, backend=backend)
+    composite = factory.CompositeSource(
+        test_source, (cell_coordinate_source, height_coordinate_source)
+    )
+    x = composite.get("foo")
+    assert isinstance(x, gtx.Field)
+    assert dims.CellDim in x.domain.dims
+    assert dims.KDim in x.domain.dims
```
Comment (nfarabullini) [inline model/common/tests/states_test/test_factory.py:185]:
```suggestion
    assert set([dims.CellDim, dims.KDim]).issubset(set([x.domain.dims]))
```
How about this?

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
launch jenkins spack

**Comment by halungge** [comment]
cscs-ci run

**Comment by halungge** [comment]
cscs-ci run default

**Pull request merged**